File: src/main/java/com/teamwizardry/wizardry/common/prism/ModuleSerializer.java
package com.teamwizardry.wizardry.common.prism;

import java.util.Map.Entry;

import com.teamwizardry.librarianlib.scribe.nbt.NbtSerializer;
import com.teamwizardry.wizardry.common.spell.component.ComponentRegistry;
import com.teamwizardry.wizardry.common.spell.component.Module;

import net.minecraft.nbt.NbtElement;
import net.minecraft.nbt.NbtString;

public class ModuleSerializer extends NbtSerializer<Module>
{
    private static final ModuleSerializer INSTANCE = new ModuleSerializer();
    private ModuleSerializer() {}
    public static ModuleSerializer get() { return INSTANCE; }
    
    @Override
    protected Module deserialize(NbtElement nbt)
    {
        return ComponentRegistry.getModules().get(nbt.asString());
    }

    @Override
    protected NbtElement serialize(Module module)
    {
        return NbtString.of(ComponentRegistry.getModules().entrySet().stream().filter(entry -> entry.getValue().equals(module)).map(Entry::getKey).findFirst().get());
    }
}

File: src/main/java/com/teamwizardry/wizardry/common/prism/ModuleSerializer.kt
package com.teamwizardry.wizardry.common.prism

import java.util.function.Function
import java.util.function.Predicate

open class ModuleSerializer private constructor() : NbtSerializer<Module?>() {
    protected override fun deserialize(nbt: NbtElement): Module {
        return ComponentRegistry.getModules().get(nbt.asString())
    }

    protected override fun serialize(module: Module): NbtElement {
        return NbtString.of(
            ComponentRegistry.getModules().entries.stream()
                .filter(Predicate { (_, value): Map.Entry<String?, com.teamwizardry.wizardry.common.spell.component.Module> -> value == module })
                .map<String>(
                    Function { (key): Map.Entry<String?, com.teamwizardry.wizardry.common.spell.component.Module?> -> key })
                .findFirst().get()
        )
    }

    companion object {
        private val INSTANCE = ModuleSerializer()
        fun get(): ModuleSerializer {
            return INSTANCE
        }
    }
}